language/move-prover/src/lib.rs

```diff
@@ -313,18 +317,50 @@ fn verification_analysis(env: &mut GlobalEnv, targets: &FunctionTargetsHolder) {
     }
 
     for module_env in env.get_modules() {
-        let module_id = module_env.get_id();
         for func_env in module_env.get_functions() {
+            if func_env.has_friend() {
+                let loc = func_env.get_loc();
+                if func_env.is_opaque() {
+                    env.error(&loc, "function with a friend cannot be declared as opaque");
+                }
+                let calling_functions = func_env.get_calling_functions();
+
+                // If a function is called by some caller, check that the caller
+                // is indeed its friend.
+                if !calling_functions.is_empty() {
+                    if let Some(friend_env) = func_env.get_friend_env() {
+                        // The function's friend is in the environment.
+                        if calling_functions.len() != 1
```

Comment:
This is interesting. We currently can have only one `pragma friend`, though (I think) the syntax allows for repeated friend declarations. This restriction is OK until we have full visibility, but perhaps we should provide an error message at the friend declaration when there are multiple ones.